cmake/pcl_find_sse.cmake

```
@@ -1,210 +1,254 @@
-        int main ()
-        {
-          __m128i a, b;
-          int vals[4] = {-1, -2, -3, -4};
-          a = _mm_loadu_si128 ((const __m128i*)vals);
-          b = _mm_abs_epi32 (a);
-          _mm_storeu_si128 ((__m128i*)vals, b);
+      #include <emmintrin.h>
+      int main ()
+      {
+          __m128d a, b;
+          double vals[2] = {0};
+          a = _mm_loadu_pd (vals);
+          b = _mm_add_pd (a,a);
+          _mm_storeu_pd (vals,b);
           return (0);
-        }"
-        HAVE_SSSE3_EXTENSIONS)
+      }"
+      HAVE_SSE2_EXTENSIONS)
 
-    if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
-        set(CMAKE_REQUIRED_FLAGS "-msse3")
+    if(HAVE_SSE2_EXTENSIONS)
+      set(SSE_LEVEL 2.0)
     endif()
+  endif()
 
-    check_cxx_source_runs("
-        #include <pmmintrin.h>
-        int main ()
-        {
-            __m128d a, b;
-            double vals[2] = {0};
-            a = _mm_loadu_pd (vals);
-            b = _mm_hadd_pd (a,a);
-            _mm_storeu_pd (vals, b);
-            return (0);
-        }"
-        HAVE_SSE3_EXTENSIONS)
-
+  if(SSE_LEVEL LESS 2.0)
     if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
-        set(CMAKE_REQUIRED_FLAGS "-msse2")
+      set(CMAKE_REQUIRED_FLAGS "-msse")
     elseif(MSVC AND NOT CMAKE_CL_64)
-        set(CMAKE_REQUIRED_FLAGS "/arch:SSE2")
+      set(CMAKE_REQUIRED_FLAGS "/arch:SSE")
     endif()
 
     check_cxx_source_runs("
-        #include <emmintrin.h>
-        int main ()
-        {
-            __m128d a, b;
-            double vals[2] = {0};
-            a = _mm_loadu_pd (vals);
-            b = _mm_add_pd (a,a);
-            _mm_storeu_pd (vals,b);
-            return (0);
-        }"
-        HAVE_SSE2_EXTENSIONS)
+      #include <xmmintrin.h>
+      int main ()
+      {
+          __m128 a, b;
+          float vals[4] = {0};
+          a = _mm_loadu_ps (vals);
+          b = a;
+          b = _mm_add_ps (a,b);
+          _mm_storeu_ps (vals,b);
+          return (0);
+      }"
+      HAVE_SSE_EXTENSIONS)
 
-    if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
-        set(CMAKE_REQUIRED_FLAGS "-msse")
-    elseif(MSVC AND NOT CMAKE_CL_64)
-        set(CMAKE_REQUIRED_FLAGS "/arch:SSE")
+    if(HAVE_SSE_EXTENSIONS)
+      set(SSE_LEVEL 1.0)
+    endif()
+  endif()
+  
+  if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
+    if(SSE_LEVEL GREATER_EQUAL 4.2)
+      set(SSE_FLAGS "-msse4.2 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 4.1)
+      set(SSE_FLAGS "-msse4.1 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 3.1)
+      set(SSE_FLAGS "-mssse3 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 3.0)
+      set(SSE_FLAGS "-msse3 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 2.0)
+      set(SSE_FLAGS "-msse2 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 1.0)
+      set(SSE_FLAGS "-msse -mfpmath=sse")
+    else()
```

Just a matter of taste.

Maybe you want to deduplicate the `-mfpmath=sse`:

```suggestion
  if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
    if(SSE_LEVEL GREATER_EQUAL 1.0)
      if(SSE_LEVEL GREATER_EQUAL 4.2)
        set(SSE_FLAGS "-msse4.2")
      elseif(SSE_LEVEL GREATER_EQUAL 4.1)
        set(SSE_FLAGS "-msse4.1")
      elseif(SSE_LEVEL GREATER_EQUAL 3.1)
        set(SSE_FLAGS "-msse3")
      elseif(SSE_LEVEL GREATER_EQUAL 3.0)
        set(SSE_FLAGS "-msse3")
      elseif(SSE_LEVEL GREATER_EQUAL 2.0)
        set(SSE_FLAGS "-msse2")
      else()
        set(SSE_FLAGS "-msse")
      endif()
      string(APPEND SSE_FLAGS " -mfpmath=sse")
    else()
```

But current variant is also okay for me.

Another possibility would be to use `set(SSE_FLAGS "-msse${SSE_LEVEL} -mfpmath=sse")` for SSE2 or greater (in this case you should change `set(SSE_LEVEL 2.0)` to `set(SSE_LEVEL 2)`). This is short, but less readable (so imho current version is maybe better).